stages/org.osbuild.uki

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/python3
+"""
+Create a Unified Kernel Image
+
+A Unified Kernel Image (UKI) is a UEFI executable that contains
+a UEFI stub loader, together with the kernel, the initrd, the
+kernel command line and (optionally) a splash screen.
+
+For more information see the specification at:
+https://github.com/uapi-group/specifications/blob/main/specs/unified_kernel_image.md
+
+Host commands: objdump, objcopy
+"""
+
+import os
+import subprocess
+import sys
+from typing import Dict, Tuple
+
+import osbuild
+from osbuild.util import pe32p
+
+SCHEMA_2 = r"""
+"options": {
+  "additionalProperties": false,
+  "required": ["filename"],
+  "properties": {
+    "filename": {
+      "description": "Filename to use for the resulting UKI",
+      "type": "string"
+    },
+    "kernel": {
+      "type": "object",
+      "required": ["opts"],
+      "properties": {
+        "opts": {
+          "description": "Array of kernel options for the UKI",
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        }
+      }
+    }
+  }
+},
+"inputs": {
+  "type": "object",
+  "additionalProperties": false,
+  "required": ["kernel", "initrd", "osrel"],
+  "properties": {
+    "osrel": {
+      "type": "object",
+      "additionalProperties": true
+    },
+    "kernel": {
+      "type": "object",
+      "additionalProperties": true
+    },
+    "initrd": {
+      "type": "object",
+      "additionalProperties": true
+    }
+  }
+}
+"""
+
+
+def align_up(address: int, alignment: int) -> int:
+    if address % alignment == 0:
+        return address
+    return ((address + alignment) // alignment) * alignment
+
+
+def detect_start_offset(stub: str) -> Tuple[int, int]:
+
+    with open(stub, "rb") as f:
+        coff = pe32p.read_coff_header(f)
+        opt = pe32p.read_optional_header(f, coff)
+        sections = pe32p.read_sections(f, coff)
+
+    assert sections, "No sections found in stub"
+
+    last = sections[-1]
+
+    offset = last.VirtualAddress + last.VirtualSize
+    alignment = max(opt.SectionAlignment, opt.FileAlignment)
```

Comment:
Why do you care for FileAlignment?

Reply:
Discussed out-of-band, added a comment to make it clear.